gremlin-go/driver/traversal_test.go

```diff
@@ -62,4 +65,349 @@ func TestTraversal(t *testing.T) {
+		// Start a transaction traversal.
+		remote := newConnection(t)
+		g := Traversal_().WithRemote(remote)
+		dropGraphCheckCount(t, g)
+
+		// Create two transactions.
+		tx1 := g.Tx()
+		tx2 := g.Tx()
+
+		// Generate two GraphTraversalSource's for each transaction with begin.
+		gtx1, _ := tx1.Begin()
+		gtx2, _ := tx2.Begin()
+		verifyTxState(t, true, tx1, tx2)
+
+		// Add stuff to both gtx.
+		addNodeValidateTransactionState(t, g, gtx1, 0, 0, tx1, tx2)
+		addNodeValidateTransactionState(t, g, gtx2, 0, 0, tx1, tx2)
+		addNodeValidateTransactionState(t, g, gtx2, 0, 1, tx1, tx2)
+		addNodeValidateTransactionState(t, g, gtx2, 0, 2, tx1, tx2)
+
+		// someone gets lazy and doesn't commit/rollback and just calls Close() but on the parent
+		// DriverRemoteConnection for all the session that were created via Tx() - the graph
+		// will decide how to treat the transaction, but for neo4j/gremlin server in this
+		// test configuration it should rollback.
+		remote.Close()
+
+		assert.False(t, tx1.IsOpen())
+		assert.False(t, tx2.IsOpen())
+		verifyGtxClosed(t, gtx1)
+		verifyGtxClosed(t, gtx2)
+
+		remote = newConnection(t)
+		g = Traversal_().WithRemote(remote)
+		assert.Equal(t, int32(0), getCount(t, g))
+	})
+}
+
+func newConnection(t *testing.T) *DriverRemoteConnection {
+	testNoAuthWithAliasUrl := getEnvOrDefaultString("GREMLIN_SERVER_URL", "ws://localhost:8182/gremlin")
+	testNoAuthWithAliasAuthInfo := &AuthInfo{}
+	testNoAuthWithAliasTlsConfig := &tls.Config{}
+
+	remote, err := NewDriverRemoteConnection(testNoAuthWithAliasUrl,
+		func(settings *DriverRemoteConnectionSettings) {
+			settings.TlsConfig = testNoAuthWithAliasTlsConfig
+			settings.AuthInfo = testNoAuthWithAliasAuthInfo
+			settings.TraversalSource = "gtx"
+		})
+	assert.Nil(t, err)
+	assert.NotNil(t, remote)
+	return remote
+}
+
+func addNodeValidateTransactionState(t *testing.T, g, gAddTo *GraphTraversalSource,
+	gStartCount, gAddToStartCount int32, txVerifyList ...*transaction) {
+	// Add a single node to gAddTo, but not g.
+	// Check that vertex count in g is gStartCount and vertex count in gAddTo is gAddToStartCount + 1.
+	addV(t, gAddTo, "lyndon")
+	assert.Equal(t, gAddToStartCount+1, getCount(t, gAddTo))
+	assert.Equal(t, gStartCount, getCount(t, g))
+	verifyTxState(t, true, txVerifyList...)
+}
+
+func verifyTxState(t *testing.T, expected bool, gtxList ...*transaction) {
+	for _, tx := range gtxList {
+		assert.Equal(t, expected, tx.IsOpen())
+	}
+}
+
+func addV(t *testing.T, g *GraphTraversalSource, name string) {
+	_, promise, err := g.AddV("person").Property("name", name).Iterate()
+	assert.Nil(t, err)
+	assert.Nil(t, <-promise)
+}
+
+func dropGraphCheckCount(t *testing.T, g *GraphTraversalSource) {
+	dropGraph(t, g)
+	assert.Equal(t, int32(0), getCount(t, g))
+}
+
+func verifyGtxClosed(t *testing.T, gtx *GraphTraversalSource) {
+	// todo: should we wait for sessionBasedConnection.Close()?
+	time.Sleep(1000 * time.Millisecond)
```

Comment:
I don't think we need to wait for a sleep if we have correct waiting done on result/close

Reply:
removed in https://github.com/lyndonbauto/tinkerpop/pull/68/commits/0942c95e9d5999d883115fd432c0a16587d7cc2a